src/smart-escrow/SmartEscrow.sol

```diff
@@ -0,0 +1,239 @@
+    event BenefactorUpdated(address oldBenefactor, address newBenefactor);
+    event ContractTerminated();
+
+    IERC20 public constant OP_TOKEN = IERC20(0x4200000000000000000000000000000000000042);
+    bytes32 public constant BENEFICIARY_OWNER = keccak256("smartescrow.roles.beneficiary");
+    bytes32 public constant BENEFACTOR_OWNER = keccak256("smartescrow.roles.benefactor");
+    bytes32 public constant ESCROW_OWNER = keccak256("smartescrow.roles.escrowowner");
+
+    address private _beneficiary;
+    address private _benefactor;
+    uint256 private _released;
+    uint64 private immutable _start;
+    uint64 private immutable _end;
+    uint64 private immutable _vestingPeriod;
+    uint64 private immutable _initialTokens;
+    uint64 private immutable  _vestingEventTokens;
+    bool private _contractTerminated;
+
+    /**
+     * @dev Set initial parameters.
+     * @param beneficiaryOwner Address which can update the beneficiary address.
+     * @param benefactorOwner Address which can update the benefactor address.
+     * @param escrowOwner Address which can terminate the contract.
+     * @param beneficiaryAddress Address which receives tokens that have vested.
+     * @param benefactorAddress Address which receives remaining tokens if contract is terminated.
+     * @param startTimestamp Timestamp of the start of vesting period (or the cliff, if there is one).
+     * @param endTimestamp Timestamp of the end of the vesting period.
+     * @param vestingPeriodSeconds Period of time between each vesting event in seconds.
+     * @param numInitialTokens Number of OP tokens which vest at start time.
+     * @param numVestingEventTokens Number of OP tokens which vest upon each vesting event.
+     */
+    constructor(
+        address beneficiaryOwner,
+        address benefactorOwner,
+        address escrowOwner,
+        address beneficiaryAddress,
+        address benefactorAddress,
+        uint64 startTimestamp,
+        uint64 endTimestamp,
+        uint64 vestingPeriodSeconds,
+        uint64 numInitialTokens,
+        uint64 numVestingEventTokens
+    ) {
+        require(beneficiaryOwner != address(0), "SmartEscrow: beneficiaryOwner is zero address");
+        require(benefactorOwner != address(0), "SmartEscrow: benefactorOwner is zero address");
+        require(escrowOwner != address(0), "SmartEscrow: escrowOwner is zero address");
+        require(beneficiaryAddress != address(0), "SmartEscrow: beneficiaryAddress is zero address");
+        require(benefactorAddress != address(0), "SmartEscrow: benefactorAddress is zero address");
+
+        require(startTimestamp > 0, "SmartEscrow: start time is zero");
+        require(endTimestamp > startTimestamp, "SmartEscrow: end time is before start time");
+        require(vestingPeriodSeconds > 0, "SmartEscrow: vestingPeriodSeconds is zero");
+
+        // TODO: could add a check that checks the owners of the escrowOwner are beneficiaryOwner and benefactorOwner
```

Comment:
Thoughts on adding this check? Im a bit double minded about how general/specific to make this contract. In theory I could see other OP chains using this contract for token distribution as well, but maybe it's worth just making it specific to this use case

Reply:
I think better to just make specific to our use case